pulsar/internal/connection.go

```diff
@@ -936,23 +936,29 @@ func (c *connection) ResetLastActive() {
 }
 
 func (c *connection) isIdle() bool {
-	c.pendingLock.Lock()
-	if len(c.pendingReqs) != 0 {
-		return false
+	{
+		c.pendingLock.Lock()
+		defer c.pendingLock.Unlock()
+		if len(c.pendingReqs) != 0 {
+			return false
+		}
 	}
-	c.pendingLock.Unlock()
 
-	c.listenersLock.RLock()
-	if len(c.listeners) != 0 {
-		return false
+	{
+		c.listenersLock.RLock()
```

Comment:
Can you explain why the previous code had a deadlock? I don't understand why it is in `{}`.

Reply:
```
c.pendingLock.Lock()
if len(c.pendingReqs) != 0 {
	return false
}
c.pendingLock.Unlock()
```

In the previous code, I forgot to unlock pendingLock when pendingReqs is not empty. So it's flaky because there may be some message in the pendingReqs when triggering the idle check.

> I don't understand why it is in {}.
Using `{}` here is to make sure we only acquire one lock at a time to avoid deadlock. So the lock will be automatically unlocked when out of this scope.